src/ssh/cockpitsshrelay.c

```diff
@@ -1320,61 +1320,67 @@ send_auth_reply (CockpitSshData *data,
   return ret;
 }
 
-static void
+static gboolean
 parse_host (const gchar *host,
             gchar **hostname,
             gchar **username,
             guint *port)
 {
-  gchar *user_arg = NULL;
-  gchar *host_arg = NULL;
-  gchar *tmp = NULL;
-  gchar *end = NULL;
-
-  guint64 tmp_num;
-
-  gsize host_offset = 0;
-  gsize host_length = strlen (host);
-
-  tmp = strrchr (host, '@');
-  if (tmp)
-    {
-      if (tmp[0] != host[0])
+  GError *error = NULL;
+  g_autoptr (GRegex) regex = g_regex_new ("^"
+         "(?:(.+)@)?"         /* optional username */
+         "(?|"                /* one of... */
+           "\\[([^]@]+)\\]"     /* hostname in square brackets, no @ */
+           "(?::([0-9]+))?"     /* optional port number */
+         "|"                  /* or */
+           "([^@:]+)"           /* hostname with no : or @ */
+           "(?::([0-9]+))?"     /* optional port number */
+         "|"                  /* or */
+           "([^@]+)"            /* hostname with no @ but : (IPv6 address), and no port */
+         ")"                  /* . */
+       "$",
+       0, 0, &error);
+  g_assert_no_error (error);
+
+  g_autoptr(GMatchInfo) info = NULL;
+
+  if (g_regex_match (regex, host, 0, &info))
+    {
+      g_autofree gchar *port_str = g_match_info_fetch (info, 3);
+      gchar *endptr;
```

Comment:
Pull those (`endptr`, `value`) inside the block below (optional: put them on the same line where they are first used).

Reply:
Done.